[BEAM-7948] Add time-based cache threshold support in the Java data s…

sunjincheng121 opened the pull request:
Currently only size-based cache threshold is supported in data service. It should also support the time-based cache threshold. This is very important, especially for streaming jobs which are sensitive to the delay. So, This PR add the time-based cache threshold support in the Java data service.


Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

lukecwik commented on sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataBufferingOutboundObserver.java:
```diff
@@ -43,48 +57,83 @@
+  private final AtomicReference<IOException> flushException = new AtomicReference<>(null);
+  private final Object lock = new Object();
 
   private BeamFnDataBufferingOutboundObserver(
```
I think we will want to have the original  implementation of this class so that we don't need to take on the synchronization overhead for flush.

lukecwik commented on sdks/java/fn-execution/src/test/java/org/apache/beam/sdk/fn/data/BeamFnDataBufferingOutboundObserverTest.java:
```diff
@@ -146,6 +159,85 @@ public void testConfiguredBufferLimit() throws Exception {
+
+    // Test that it emits when time passed the time limit
+    consumer.accept(valueInGlobalWindow(new byte[1]));
+    Thread.sleep(10); // allow the flush thread to flush the buffer
```
This test will be flaky since its based upon very small amounts of time. You want to make withOnNext free a lock/semaphore/countdownlatch and in your test you wait on it with a large time limit instead of using thread.sleep

lukecwik commented on sdks/java/fn-execution/src/test/java/org/apache/beam/sdk/fn/data/BeamFnDataBufferingOutboundObserverTest.java:
```diff
@@ -146,6 +159,85 @@ public void testConfiguredBufferLimit() throws Exception {
+
+    // Test that it emits when time passed the time limit
+    consumer.accept(valueInGlobalWindow(new byte[1]));
+    Thread.sleep(100); // allow the flush thread to flush the buffer
```
ditto here to use a semaphore/lock/countdownlatch instead of thread.sleep

sunjincheng121 commented:
@lukecwik  Thanks a lot for your review. I have updated the PR and addressed your comments.

> I was wondering how this manifests?
> 
> Do you have a way to cancel downstream computations based upon the output of this bundle if this bundle fails?

If a bundle fails, the exception will be prograted to the main thread at the next time when an input element is processed(accept/close is called). So I think the failure handle logic for time-based data buffer strategy is the same as the size-based data buffer strategy. Time-based data buffer strategy has not and also should not introduce extra complexity for failure handling.

Looking forward to your feedback.

sunjincheng121 commented:
Hi @lukecwik, Could you please have another look, any comment is welcome, Thanks! :)

lukecwik commented on sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataTimeBasedBufferingOutboundObserver.java:
```diff
@@ -0,0 +1,109 @@
+    super.accept(t);
+  }
+
+  private void periodicFlush() {
```
I don't think you need flushException or this method at all, you can invoke `this::flush` from the scheduled executor directly and check the status of the flushFuture via the `isDone()` method. If its done then you know that an exception was thrown.

sunjincheng121 replied:
It's not possible to use the flush method directly as it throws an IOException. It's a good idea to use `flushFuture.isDone()` to check the status of the flush thread. Will be Updated. :)

lukecwik replied:
nit:
You can use the callable version of schedule to be able to use methods that throw exceptions:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html#schedule-java.util.concurrent.Callable-long-java.util.concurrent.TimeUnit-

lukecwik commented on sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataTimeBasedBufferingOutboundObserver.java:
```diff
@@ -0,0 +1,109 @@
+
+  /** Check if the flush thread failed with an exception. */
+  private void checkFlushThreadException() throws IOException {
+    IOException e = flushException.get();
```
```
if (flushFuture.isDone()) {
  try {
    flushFuture.get();
    throw new IOException("Periodic flushing thread finished unexpectedly.");
  } catch (ExecutionException e) {
    if (e.getCause() instanceof IOException) {
      throw (IOException) e.getCause();
    }
    throw new RuntimeException(e);
  }
}
```

lukecwik commented on sdks/java/fn-execution/src/test/java/org/apache/beam/sdk/fn/data/BeamFnDataTimeBasedBufferingOutboundObserverTest.java:
```diff
@@ -0,0 +1,139 @@
+      // expected
+    }
+
+    consumer =
```
It seems like your duplicating the above portion of the test here unless I'm not seeing what the small difference is.

sunjincheng121 replied:
The first test case verifies that the exception is propagated when processing the next element. The second test case verifies that the exception is propagated when closing.

lukecwik commented on sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataTimeBasedBufferingOutboundObserver.java:
```diff
@@ -0,0 +1,109 @@
+      try {
+        flushFuture.get();
+      } catch (CancellationException | ExecutionException | InterruptedException exn) {
+        // expected
```
CancellationException is the only exception that is expected. ExecutionException should be unwrapped and we should throw the cause (if set) otherwise we should throw the execution exception directly. InterruptedException should not be caught here at all.

sunjincheng121 commented on sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataTimeBasedBufferingOutboundObserver.java:
```diff
@@ -0,0 +1,109 @@
+  public void close() throws Exception {
+    checkFlushThreadException();
+    synchronized (lock) {
+      flushFuture.cancel(true);
```
This is done by holding the flush lock and so it's guaranteed that the flush thread is not flush data. I have renamed the "lock" to "flushLock" to make it more clear.

lukecwik replied:
You wouldn't need to have the flushLock object at all if you marked the flush method as synchronized and decided cancel the flushFuture without interrupting it.

lukecwik commented:
Run Java PreCommit

lukecwik commented on sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataTimeBasedBufferingOutboundObserver.java:
```diff
@@ -90,20 +90,27 @@ private void periodicFlush() {
+        }
+        throw re;
+      } catch (InterruptedException | CancellationException e) {
+        // Should never happen
```
Prefer to wrap the InerruptedException and CancellationException in an IOException.

Also for the interrupted exception you should invoke Thread.interrupt() before rethrowing.

lukecwik commented on sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataTimeBasedBufferingOutboundObserver.java:
```diff
@@ -48,25 +46,27 @@
                     .setDaemon(true)
                     .setNameFormat("DataBufferOutboundFlusher-thread")
                     .build())
             .scheduleAtFixedRate(this::periodicFlush, timeLimit, timeLimit, TimeUnit.MILLISECONDS);
```
You should be able to use the callable version of schedule:
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ScheduledExecutorService.html#schedule-java.util.concurrent.Callable-long-java.util.concurrent.TimeUnit-
to invoke the flush method. You might need to provide a cast to Callable<Void> to tell Java to use that version of schedule.

sunjincheng121 replied:
I found that the callable version of schedule is not a periodic action. So, we have to create a new method which warps the `flush`, What do you think?

lukecwik replied:
Your right, I missed the fact that there was no periodic schedule that took a callable.

sunjincheng121 commented:
I have update the PR accordingly except one comment I am not pretty sure, and left comment to you. :) Appreciate if you can have another look :) @lukecwik

lukecwik commented on sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataSizeBasedBufferingOutboundObserver.java:
```diff
@@ -85,7 +85,7 @@ public void close() throws Exception {
 
   @Override
-  public void flush() throws IOException {
+  public synchronized void flush() throws IOException {
```
We don't want to make this synchronized since this class is not thread safe and should not take this perf hit. Your previous usage of defining an override that is synchronized for the class that called super.flush() is all you need.

lukecwik commented:
Just remove the synchronized from BeamFnDataSizeBasedBufferingOutboundObserver.java and add
```
@Override
public synchronized void flush() throws IOException {
  super.flush();
}
```
to the time based outbound observer and we can merge.

sunjincheng121 commented:
Thanks for the review @lukecwik 👍
